quarkchain/cluster/master.py

```diff
@@ -1017,13 +1033,15 @@ def handle_new_root_block_header(self, header, peer):
         All update root block should be done in serial to avoid inconsistent global root block state.
         """
         self.root_state.validate_block(r_block)  # throw exception if failed
-        update_tip = False
         try:
             update_tip = self.root_state.add_block(r_block)
-            success = True
         except ValueError:
             Logger.log_exception()
-            success = False
+            return
+
+        # root block is written to db but not broadcasted to slaves
+        # use write-ahead log so if crashed the root block can be re-broadcasted
+        self.root_state.write_committing_hash(r_block.header.get_hash())
```

Comment:
I guess we need to put the code before add_block - if shutdown happens between add_block and write_committing_hash, and restarting the node won't recover the correct block

Reply:
on the other hand, after changing the order, if it crashes between writing committing hash and the root block, we won't find the root block pointed by the committing hash - which should be fine since it's not broadcasted yet, thus can safely ignore.

good find